Archiver/src/EngineEnvironment.cpp

```diff
@@ -0,0 +1,192 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  In no event and under no legal theory, whether in tort (including negligence),
+ *  contract, or otherwise, unless required by applicable law (such as deliberate
+ *  and grossly negligent acts) or agreed to in writing, shall any Contributor be
+ *  liable for any damages, including any direct, indirect, special, incidental,
+ *  or consequential damages of any character arising as a result of this License or
+ *  out of the use or inability to use the software (including but not limited to damages
+ *  for loss of goodwill, work stoppage, computer failure or malfunction, or any and
+ *  all other commercial damages or losses), even if such Contributor has been advised
+ *  of the possibility of such damages.
+ */
+
+
+#include "pch.h"
+#include "argparse.hpp"
+#include <fstream>
+
+namespace Diligent
+{
+
+EngineEnvironment* EngineEnvironment::m_pEnvironment = nullptr;
+
+IArchiverFactory* EngineEnvironment::GetArchiveFactory()
+{
+    return m_pArchiveBuilderFactory;
+}
+
+ISerializationDevice* EngineEnvironment::GetSerializationDevice()
+{
+    return m_pSerializationDevice;
+}
+
+IShaderSourceInputStreamFactory* EngineEnvironment::GetShaderSourceInputStreamFactory()
+{
+    return m_pShaderStreamFactory;
+}
+
+DeviceObjectReflection* EngineEnvironment::GetDeviceObjectReflection()
+{
+    return m_pDeviceReflection.get();
+}
+
+const EngineEnvironment::EngineEnvironmentDesc& EngineEnvironment::GetDesc() const
+{
+    return m_Desc;
+}
+
+EngineEnvironment::EngineEnvironment(const EngineEnvironmentDesc& Desc) :
+    m_Desc(Desc)
+{
+
+#if EXPLICITLY_LOAD_ARCHIVER_FACTORY_DLL
+    auto GetArchiverFactory = LoadArchiverFactory();
+    if (GetArchiverFactory != nullptr)
+        m_pArchiveBuilderFactory = GetArchiverFactory();
+#else
+    m_pArchiveBuilderFactory = Diligent::GetArchiverFactory();
+#endif
+
+    SerializationDeviceCreateInfo DeviceCI       = {};
+    auto                          pAllocatorHack = std::make_unique<DeviceObjectReflection>(RefCntAutoPtr<ISerializationDevice>(nullptr), RefCntAutoPtr<IShaderSourceInputStreamFactory>(nullptr), RENDER_DEVICE_TYPE_FLAG_NONE);
+
+    if (!Desc.ConfigFilePath.empty())
+    {
+        std::ifstream  stream(Desc.ConfigFilePath);
+        nlohmann::json Json = nlohmann::json::parse(stream);
+        Deserialize(Json, DeviceCI, pAllocatorHack.get());
+    }
+
+    m_pArchiveBuilderFactory->CreateSerializationDevice(DeviceCI, &m_pSerializationDevice);
+    m_pArchiveBuilderFactory->CreateDefaultShaderSourceStreamFactory(m_Desc.ShadersFilePath.c_str(), &m_pShaderStreamFactory);
+    m_pDeviceReflection = std::make_unique<DeviceObjectReflection>(m_pSerializationDevice, m_pShaderStreamFactory, m_Desc.DeviceBits);
+}
+
+EngineEnvironment::~EngineEnvironment()
+{}
+
+void EngineEnvironment::Initialize(int argc, char** argv)
+{
+    argparse::ArgumentParser Parser{"JSON Archiver"};
```

Comment:
Generally, Archiver should everywhere be replaced with either Packager or Parser.
Where it reads or write data from JSON, it is Parser.